### PR TITLE
[SessionD] Schedule bearer creation for Session Create

### PR DESCRIPTION
**Pull request by themarwhal**
Signed-off-by: Marie Bremner <marwhal@fb.com>

## Summary
This PR adds a change to delay dedicated bearer creation by a configured amount of time for session creation. This is needed since, we need to send the bearer request after UE attach is completed.

When we process the create session response, if a bearer creation is needed we schedule the request to be sent in the event loop. When the function is called, we double check that the request is still valid before making the request. The conditions we check are:
1. The session is exists
2. The session is active
3. The rule exists in the session
4. A bearer creation is still needed -> meaning that the rule still has a qos / is not already tied to a bearer etc.

This amount is configured in `sessiond.yml` under : `bearer_creation_delay_on_session_init`. 

## Test Plan
Additional SessionD unit tests to cover the edge cases
SessionD unit tests + S1AP tests
Also did some manual testing by modifying the dynamic rule PolicyDB returns on SessionCreation to include a QoS field and running an `attach-detach` test. I've set ` bearer_creation_delay_on_session_init: 1000`
```
def get_allow_all_policy_rule(
    subscriber_id: str,
    apn: str,
) -> PolicyRule:
    policy_id = _get_allow_all_rule_id(subscriber_id, apn)
    return PolicyRule(
        id=policy_id,
        priority=2,
        flow_list=_get_allow_all_flows(),
        tracking_type=PolicyRule.TrackingType.Value("NO_TRACKING"),
        qos = FlowQos(
            qci=5,  # qci value [1 to 9]
            max_req_bw_ul=10000000,  # MAX bw Uplink
            max_req_bw_dl=15000000,  # MAX bw Downlink
            gbr_ul=1000000,  # GBR Uplink
            gbr_dl=2000000,  # GBR Downlink
            arp=QosArp(
                priority_level=15,  # ARP priority
                pre_capability=1,  # pre-emption capability
                pre_vulnerability=1,  # pre-emption vulnerability
            ),
        ),
    )
```
Here is the logs from SessionD, the important lines are "Scheduling a bearer creation request for newly created IMSI001010000000001-623138 in 1000 ms" @ Sep 11 12:38:48 and "Creating dedicated bearers for IMSI001010000000001, 192.168.128.13, rules={ allowlist_sid-IMSI001010000000001-magma.ipv4 }" @ Sep 11 12:38:49.

```Sep 11 12:38:48 magma-dev sessiond[12503]: I0911 12:38:48.831861 12503 LocalSessionManagerHandler.cpp:583] Received a LocalCreateSessionRequest for IMSI001010000000001 with APN:magma.ipv4, default bearer ID:5, PLMN ID:00101, IMSI PLMN ID:00101
Sep 11 12:38:48 magma-dev sessiond[12503]: I0911 12:38:48.831985 12503 LocalSessionManagerHandler.cpp:234] Sending a CreateSessionRequest to fetch policies for IMSI001010000000001-623138
Sep 11 12:38:48 magma-dev sessiond[12503]: I0911 12:38:48.834321 12503 GrpcMagmaUtils.cpp:37]
Sep 11 12:38:48 magma-dev sessiond[12503]:   magma.lte.CreateSessionResponse {
Sep 11 12:38:48 magma-dev sessiond[12503]:       dynamic_rules {
Sep 11 12:38:48 magma-dev sessiond[12503]:         policy_rule {
Sep 11 12:38:48 magma-dev sessiond[12503]:           id: "allowlist_sid-IMSI001010000000001-magma.ipv4"
Sep 11 12:38:48 magma-dev sessiond[12503]:           priority: 2
Sep 11 12:38:48 magma-dev sessiond[12503]:           flow_list {
Sep 11 12:38:48 magma-dev sessiond[12503]:             match {
Sep 11 12:38:48 magma-dev sessiond[12503]:             }
Sep 11 12:38:48 magma-dev sessiond[12503]:           }
Sep 11 12:38:48 magma-dev sessiond[12503]:           flow_list {
Sep 11 12:38:48 magma-dev sessiond[12503]:             match {
Sep 11 12:38:48 magma-dev sessiond[12503]:               direction: DOWNLINK
Sep 11 12:38:48 magma-dev sessiond[12503]:             }
Sep 11 12:38:48 magma-dev sessiond[12503]:           }
Sep 11 12:38:48 magma-dev sessiond[12503]:           qos {
Sep 11 12:38:48 magma-dev sessiond[12503]:             max_req_bw_ul: 10000000
Sep 11 12:38:48 magma-dev sessiond[12503]:             max_req_bw_dl: 15000000
Sep 11 12:38:48 magma-dev sessiond[12503]:             gbr_ul: 1000000
Sep 11 12:38:48 magma-dev sessiond[12503]:             gbr_dl: 2000000
Sep 11 12:38:48 magma-dev sessiond[12503]:             qci: QCI_5
Sep 11 12:38:48 magma-dev sessiond[12503]:             arp {
Sep 11 12:38:48 magma-dev sessiond[12503]:               priority_level: 15
Sep 11 12:38:48 magma-dev sessiond[12503]:               pre_capability: PRE_CAP_DISABLED
Sep 11 12:38:48 magma-dev sessiond[12503]:               pre_vulnerability: PRE_VUL_DISABLED
Sep 11 12:38:48 magma-dev sessiond[12503]:             }
Sep 11 12:38:48 magma-dev sessiond[12503]:           }
Sep 11 12:38:48 magma-dev sessiond[12503]:           tracking_type: NO_TRACKING
Sep 11 12:38:48 magma-dev sessiond[12503]:         }
Sep 11 12:38:48 magma-dev sessiond[12503]:       }
Sep 11 12:38:48 magma-dev sessiond[12503]:       session_id: "IMSI001010000000001-623138"
Sep 11 12:38:48 magma-dev sessiond[12503]:   }
Sep 11 12:38:48 magma-dev sessiond[12503]: I0911 12:38:48.834350 12503 LocalSessionManagerHandler.cpp:244] Received a CreateSessionResponse for IMSI001010000000001-623138
Sep 11 12:38:48 magma-dev sessiond[12503]: I0911 12:38:48.834468 12503 LocalEnforcer.cpp:695] Activating untracked rule allowlist_sid-IMSI001010000000001-magma.ipv4
Sep 11 12:38:48 magma-dev sessiond[12503]: I0911 12:38:48.834828 12503 PipelinedClient.cpp:54] Sending AMBR info for IMSI001010000000001, ip addr=192.168.128.13, dl=100000000, ul=200000000
Sep 11 12:38:48 magma-dev sessiond[12503]: I0911 12:38:48.835069 12503 GrpcMagmaUtils.cpp:37]
Sep 11 12:38:48 magma-dev sessiond[12503]:   magma.lte.ActivateFlowsRequest {
Sep 11 12:38:48 magma-dev sessiond[12503]:       sid {
Sep 11 12:38:48 magma-dev sessiond[12503]:         id: "IMSI001010000000001"
Sep 11 12:38:48 magma-dev sessiond[12503]:       }
Sep 11 12:38:48 magma-dev sessiond[12503]:       ip_addr: "192.168.128.13"
Sep 11 12:38:48 magma-dev sessiond[12503]:       dynamic_rules {
Sep 11 12:38:48 magma-dev sessiond[12503]:         id: "allowlist_sid-IMSI001010000000001-magma.ipv4"
Sep 11 12:38:48 magma-dev sessiond[12503]:         priority: 2
Sep 11 12:38:48 magma-dev eventd[12819]: DEBUG:root:Logging event: stream_name: "sessiond"
Sep 11 12:38:48 magma-dev sessiond[12503]:         flow_list {
Sep 11 12:38:48 magma-dev sessiond[12503]:           match {
Sep 11 12:38:48 magma-dev sessiond[12503]:           }
Sep 11 12:38:48 magma-dev sessiond[12503]:         }
Sep 11 12:38:48 magma-dev sessiond[12503]:         flow_list {
Sep 11 12:38:48 magma-dev sessiond[12503]:           match {
Sep 11 12:38:48 magma-dev sessiond[12503]:             direction: DOWNLINK
Sep 11 12:38:48 magma-dev sessiond[12503]:           }
Sep 11 12:38:48 magma-dev sessiond[12503]:         }
Sep 11 12:38:48 magma-dev sessiond[12503]:         qos {
Sep 11 12:38:48 magma-dev sessiond[12503]:           max_req_bw_ul: 10000000
Sep 11 12:38:48 magma-dev sessiond[12503]:           max_req_bw_dl: 15000000
Sep 11 12:38:48 magma-dev sessiond[12503]:           gbr_ul: 1000000
Sep 11 12:38:48 magma-dev sessiond[12503]:           gbr_dl: 2000000
Sep 11 12:38:48 magma-dev sessiond[12503]:           qci: QCI_5
Sep 11 12:38:48 magma-dev sessiond[12503]:           arp {
Sep 11 12:38:48 magma-dev sessiond[12503]:             priority_level: 15
Sep 11 12:38:48 magma-dev sessiond[12503]:             pre_capability: PRE_CAP_DISABLED
Sep 11 12:38:48 magma-dev sessiond[12503]:             pre_vulnerability: PRE_VUL_DISABLED
Sep 11 12:38:48 magma-dev sessiond[12503]:           }
Sep 11 12:38:48 magma-dev sessiond[12503]:         }
Sep 11 12:38:48 magma-dev sessiond[12503]:         tracking_type: NO_TRACKING
Sep 11 12:38:48 magma-dev sessiond[12503]:       }
Sep 11 12:38:48 magma-dev sessiond[12503]:       request_origin {
Sep 11 12:38:48 magma-dev sessiond[12503]:       }
Sep 11 12:38:48 magma-dev sessiond[12503]:       apn_ambr {
Sep 11 12:38:48 magma-dev sessiond[12503]:         max_bandwidth_ul: 200000000
Sep 11 12:38:48 magma-dev sessiond[12503]:         max_bandwidth_dl: 100000000
Sep 11 12:38:48 magma-dev sessiond[12503]:       }
Sep 11 12:38:48 magma-dev sessiond[12503]:   }
Sep 11 12:38:48 magma-dev sessiond[12503]: I0911 12:38:48.835448 12503 LocalEnforcer.cpp:933] Scheduling a bearer creation request for newly created IMSI001010000000001-623138 in 1000 ms
Sep 11 12:38:48 magma-dev sessiond[12503]: I0911 12:38:48.836154 12503 LocalSessionManagerHandler.cpp:251] Successfully initialized new session IMSI001010000000001-623138 in SessionD for subscriber IMSI001010000000001
Sep 11 12:38:49 magma-dev sessiond[12503]: I0911 12:38:49.836793 12503 SpgwServiceClient.cpp:107] Creating dedicated bearers for IMSI001010000000001, 192.168.128.13, rules={ allowlist_sid-IMSI001010000000001-magma.ipv4 }
```

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Review by ulaskozat — APPROVED** [review]

**Review by ardzoht — APPROVED** [review]

File: lte/gateway/c/session_manager/LocalEnforcer.cpp [hunk 1]
```diff
@@ -896,35 +896,102 @@ void LocalEnforcer::handle_session_init_rule_updates(
...
+      [this, imsi, session_id, bearer_updates]() mutable {
+        auto session_map = session_store_.read_sessions({imsi});
+        auto it          = session_map.find(imsi);
+        if (it == session_map.end()) {
+          MLOG(MWARNING) << "Ignoring dedicated bearer creations from session "
+                            "creation for "
+                         << session_id << " since it no longer exists";
+          return;
+        }
+        for (auto& session : it->second) {
```
Comment (uri200) [inline lte/gateway/c/session_manager/LocalEnforcer.cpp:949]: what about creating a function on sessionState called `findSession(SessionMap& sm, std:string sessionId)`. We can reuse it in other places, plus you will be able to log  it here

if sessionmap were a struct we could even attach tat function to the struct 


```
		SessionMap sm = it->second;
       if (it == session_map.end() || !findSession(sm, sessionId) {
          MLOG(MWARNING) << "Ignoring dedicated bearer creations from session "
                            "creation for "
                         << session_id << " since it no longer exists";
          return;
        }

```

**Review by uri200 — APPROVED** [review]

File: lte/gateway/c/session_manager/LocalEnforcer.cpp (hunk 1 above)
Comment (themarwhal) [reply lte/gateway/c/session_manager/LocalEnforcer.cpp:949]: Yeah we have many of this use cases, we can have a function that returns an optional of a pointer to a session state or something like that. Will do this in a different PR though.
